README
Hax!

  alias t=./t
  source completion  # zsh only

  t add some task    # Add to list
  t bump 0 s<TAB>    # Move task to TODAY
  t a 0 another task # New task on today list, shortcut letter
  t list             # Today's tasks
  t done a<TAB>      # It is done!
  t b 1w s<TAB>      # Actually we will to it next week
  t l all            # All tasks


HISTORY
0.2.2 - 4 February 2011
  * rubygems-test compat

0.2.1 - 3 December 2010
  * Easier to read formatting for list
  * Can not add a blank task

0.2.0 - 28 November 2010
  * Added ZSH completion
  * Added XTDO_PATH environment variable for customizing storage location
  * Added recurring tasks to list all

0.1.0 - 27 November 2010
  * Initial release


spec/add_spec.rb
require 'spec_helper'

feature 'add' do
  scenario 'to next' do
    t('a T1')
    t('l a').should have_task('T1', :in => :next)
  end

  scenario 'to today' do
    t('a 0 T1')
    t('l a').should have_task('T1', :in => :today)
  end

  scenario 'to scheduled' do
    t('a 1w T1')
    t('l a').should have_task('T1', :in => :scheduled, :for => Date.today + 7)
  end

  scenario 'error' do
    t('a').should == "Invalid command"
  end
end


spec/bump_spec.rb
require 'spec_helper'

feature 'bumping tasks' do
  scenario 'moving around' do
    t('a T1')
    t('l').should_not have_task('T1', :in => :today)
    t('b 0 T1')
    t('l').should have_task('T1', :in => :today)
    t('b 1 T1')
    t('l a').should have_task('T1', :in => :scheduled, :for => Date.today + 1)
    t('b 1w T1')
    t('l a').should have_task('T1', :in => :scheduled, :for => Date.today + 7)
  end

  scenario 'invalid input' do
    t('b 0 T1').should == 'No such task'
    t('b T1').should == 'Invalid time'
  end
end


spec/done_spec.rb
require 'spec_helper'

feature 'done' do
  scenario 'marking off tasks' do
    t('a T1')
    t('l a').should have_task('T1')
    t('d T1').should == "Task done"
    t('l').should_not have_task('T1')
  end

  scenario 'invalid input' do
    t('d T1').should == "No such task"
  end
end


prototype.rb
require 'yaml'
require 'date'

tasks = if File.exists?("tasks.yml")
  YAML.load(File.open("tasks.yml").read)
else
  {}
end


RELATIVE_REGEX = /^(\d+)([dwmy])?$/

def parse_relative_time(str)
  number   = str[RELATIVE_REGEX, 1].to_i
  modifier = str[RELATIVE_REGEX, 2]
  adj = {
    'd' => 1,
    'w' => 7,
    'm' => 30,
    'y' => 365
  }[modifier] || 1
  Date.today + adj * number
end

case ARGV[0]
when 'add', 'a' then 
  task = ARGV[1..-1]
  if task[0] =~ RELATIVE_REGEX
    scheduled_for = parse_relative_time(task.shift)
    puts "Adding task: #{task.join(' ')} for #{scheduled_for}"
    tasks[task.join(' ')] = {:scheduled => scheduled_for} # TODO: if already exists?
  else
    puts "Adding task: #{task.join(' ')}"
    tasks[task.join(' ')] = {} # TODO: if already exists?
  end
when 'done', 'd' then tasks.delete_if {|name, _| name == ARGV[1..-1].join(" ").tr("_", " ")}
when 'bump', 'b' then
  task = ARGV[1..-1]
  scheduled_for = parse_relative_time(task.shift)
  tasks[task.join(' ').tr('_', ' ')][:scheduled] = scheduled_for
when 'bumpc' then # For completion
  puts tasks.keys.map {|x| x.tr(' ', "_") }.sort
when 'list', 'l' then
  today = Date.today
  today_tasks = tasks.select {|name, attrs| attrs[:scheduled] && attrs[:scheduled] <= today }
  puts "==== TODAY ====="
  puts today_tasks.keys.sort

  if ARGV[1] == 'all' || ARGV[1] == 'a'
    puts
    puts "===== NEXT ====="
    next_tasks = tasks.select {|name, attrs| !attrs[:scheduled] }
    puts next_tasks.keys.sort
    puts
    puts "===== SCHEDULED ====="
    scheduled_tasks = tasks.select {|name, attrs| attrs[:scheduled] }.delete_if{|k, v| today_tasks.keys.include?(k)}
    puts scheduled_tasks.sort_by{|name, attrs| attrs[:scheduled] }.map {|name, attrs| "#{attrs[:scheduled]}\t#{name}"}.join("\n")
  end
#   puts
# 
#   puts tasks.map {|name, attrs| "#{name}\t#{attrs[:scheduled]}"}.join("\n")
end

# TODO: Two step write incase of crash
File.open("tasks.yml", "w") do |f|
  f.write tasks.to_yaml
end


lib/xtdo.rb
require 'date'
require 'yaml'

class Xtdo
  def self.run(store, operation)
    operation = operation.split(/\s+/)
    verb = operation.shift

#     tasks = {
#       'T1' => {:scheduled => Date.today}
#     }
    store = File.expand_path(store)
    tasks = if File.exists?(store)
      YAML.load(File.open(store))
    else
      {}
    end
    tasks[:tasks] ||= {}
    tasks[:recurring] ||= {}

    manager = Xtdo.new(tasks)

    ret = case verb
    when 'a' then # A is for add!
      manager.add operation.join(' ')
    when 'b' then # B is for bump!
      manager.bump operation.join(' ')
    when 'd' then # D is for done!
      manager.done operation.join(' ')
    when 'l' then # L is for list!
      case operation[0]
      when 'a' then
        manager.list [:today, :next, :scheduled]
      when 'c' then
        manager.list [:today, :next, :scheduled], :format => :completion
      else
        manager.list [:today]
      end
    when 'r' then # R is for recur!
      manager.recur operation.join(' ')
    end

    manager.save(store)
    ret
  end

  attr_reader :tasks, :recurring

  def initialize(tasks)
    @tasks = tasks[:tasks]
    @recurring = tasks[:recurring]
  end

  def parse_relative_time(number, period)
    adj = {
      'd' => 1,
      'w' => 7,
      'm' => 30,
      'y' => 365
    }[period] || 1
    Date.today + adj * number
  end

  def add(task)
    match = /^(?:(\d+)([dwmy])? )?(.+)/.match(task)
    if match
      number, period, task = match.captures
      @tasks[make_key task] = {:name => task}
      @tasks[make_key task][:scheduled] = parse_relative_time(number.to_i, period) if number
    else
      "Invalid command"
    end
  end

  def bump(task)
    number, period, task = /^(?:(\d+)([dwmy])? )?(.*)/.match(task).captures
    if !number
      "Invalid time"
    elsif tasks[make_key task]
      tasks[make_key task][:scheduled] = parse_relative_time(number.to_i, period)
    else
      "No such task"
    end
  end

  def done(task)
    if tasks.delete(make_key task)
      "Task done"
    else
      "No such task"
    end
  end

  def list(groups, opts = {})
    # Check for recurring
    recurring.each do |name, task|
      if task[:next] <= Date.today
        tasks[make_key name] = {:scheduled => Date.today, :name => task[:name] }
        number, period, start, _ = self.class.extract_recur_tokens(task[:period] + ' ' + name)

        task[:next] = self.class.calculate_starting_day(Date.today, number, period, start)
      end
    end

    if opts[:format] == :completion
      tasks.keys.join "\n"
    else
      # Print groups
      task_selector = {
        :today     => lambda {|x| x && x <= Date.today },
        :next      => lambda {|x| !x },
        :scheduled => lambda {|x| x && x > Date.today }
      }

      groups.map do |group|
        t = tasks.select {|name, opts| task_selector[group][opts[:scheduled]] }
        next if t.empty?
        "\n\e[43;30m===== #{group.to_s.upcase} =====\e[0m\n\n" + t.map { |name, attrs|
          "  #{attrs[:name]}"
        }.join("\n")
      end.join("\n") + "\n\n"
    end
  end

  def recur(task)
    tokens = task.split(/\s+/)
    verb = tokens.shift
    task = tokens.join(' ')
    case verb
    when 'a' then
      number, period, start, name = self.class.extract_recur_tokens(task)

      if name
        period_string = "#{number}#{period}"
        period_string += ",#{start}" if start
        recurring[make_key name] = {
          :name   => name,
          :next   => Date.today + 1,
          :period => period_string
        }
      else
        "Invalid command"
      end
    when 'd' then
      if recurring.delete make_key(task)
        "Recurring task removed"
      else
        "No such recurring task"
      end
    when 'l' then
      "\n\e[43;30m===== RECURRING =====\e[0m\n\n" + recurring.map do |name, task|
        "  %-8s%s" % [task[:period], task[:name]]
      end.join("\n") + "\n\n"
    when 'c' then
      recurring.keys.join "\n"
    end
  end

  def make_key(name)
    name.gsub /[^a-zA-Z0-9,.]/, '-' 
  end

  def self.calculate_starting_day(date, number, period, start = nil)
    days = %w(sun mon tue wed thu fri sat)
    number = (number || 1).to_i

    adjust = case period
    when 'd' then 1
    when 'w' then
      if days.index(start)
        (days.index(start) - date.wday) % 7 + (number - 1) * 7
      end
    when 'm' then
      start = start.to_i
      if start > 0
        year = date.year
        if start.to_i <= date.day
          month = date.month + 1
          if month > 12
            month = 1
            year += 1
          end
        else
          month = date.month
        end
        Date.new(year, month, start.to_i) - date
      end
    end
    if adjust
      if adjust == 0
        date + 7
      else
        date + adjust
      end
    end
  end

  def self.extract_recur_tokens(task)
    match = /^(\d+)([dwmy])?(?:,(#{days.join('|')}|\d{1,2}))? (.*)/.match(task)
    if match
      match.captures
    else
      []
    end
  end

  def self.days
    days = %w(sun mon tue wed thu fri sat)
  end

  def save(file)
    File.open(file, 'w') {|f| f.write({
      :tasks     => tasks,
      :recurring => recurring,
      :version   => 1
    }.to_yaml) } 
  end
end
